### DISCUSSION TITLE
rust emscripten + threads, can this work?

### DISCUSSION THREAD

**Discussion by dobkeratops** [emscripten-core/emscripten#18156] [General]
emscripten is less popular in the rust community , but I use it because I like the common base between C++ and Rust of SDL2 and OpenGL bindings.

I was unfortunately unable to get *threads* working in this configuration . 

Any idea why this might be ?

There will be multiple C++ successor languages (carbon, Zig, JAI for gamedev) and of course C++ - I support continuing with C FFI interfaces which is why I have preferred to stick with Emscripten for my rust WASM project.

This comes from my '.cargo/config.toml' which specifies command line arguments passed to rustc during rusts build process. e.g .-"C" means the next string is passed as a sommandline arg.   "link-arg" in turn specifies what is passed to the linker
The commented out parts show how I was attempting to get threads working.
Other switches like 'USE_SDL=2", worked fine.

```
[target.wasm32-unknown-emscripten]
rustflags = [

#### OPTIONS FOR THREADING , WE CAN'T GET THIS WORKING SADLY #####
#
# NOTE - server needs CrossOriginHeader somethign somethign to allow SharedArrayBuffer in the browser for any of this.
#
	#"-C", "link-arg=-pthread",
	#"-C", "link-arg=-s", "-C", "link-args=USE_PTHREADS=1",
	#"-C", "target-feature=+atomics,+bulk-memory,+mutable-globals",
	#"-C", "link-arg=-s", "-C", "link-arg=WASMWORKERS",

	"-C", "link-arg=cppstuff.o",
	"-C", "link-arg=-lwebsocket.js",

	"-C", "link-arg=-s", "-C", "link-arg=WASM=1",
	"-C", "link-arg=-s", "-C", "link-arg=FORCE_FILESYSTEM=1",
	"-C", "link-arg=-s", "-C", "link-arg=USE_SDL=2",
	"-C", "link-arg=-s", "-C", "link-arg=WEBSOCKET_URL",
	"-C", "link-arg=-s", "-C", "link-arg=USE_SDL_IMAGE=2",
 	"-C", "link-arg=-s", "-C", "link-args=MAX_WEBGL_VERSION=2",
 	"-C", "link-arg=-s", "-C", "link-args=MIN_WEBGL_VERSION=2",
 	"-C", "link-arg=-s", "-C", "link-args=FETCH=1",
 	"-C", "link-arg=-s", "-C", "link-args=ASYNCIFY=1",
 	"-C", "link-arg=-s", "-C", "link-args=TOTAL_MEMORY=1024MB",
	"-C", "link-arg=-s", "-C", "link-args=PTHREAD_POOL_SIZE=4",

	"-C", "link-arg=-s", "-C", "link-args=DYNAMIC_EXECUTION=1",


#	"-C", "link-arg=-s", "-C", "link-args=DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=$readAsmConstArgs",

####### for debug symbols #################
 	"-C", "link-arg=-g",
]

```

**Comment by kripken** [comment]
I expect this might not work, since as you said it's not commonly used. But what error did you run into? Perhaps it's something fixable.

**Reply by dobkeratops** [reply]
I must admit I haven't compiled the emscripten version in a while and it'll take me time to get back into that to re-test it - I think I tried a minimal example aswell.
when it fails, it doesn't tend to give any feedback, the environment just locks up.

I think it just failed to start up. Along the way I did have to fix this 'cross origin' something issue in the page itself  to enable the shared buffers it needed (not emscripten's problem , I just want to confirm I did do that)

If I get around to compiling it again soon , I'll try to give more info.

Is there a minimal example confirming that it does actually work in C++? 
In native land rust can compile libs that C++ links to and calls of course.

I always thought it should also be possible to have a wasm C++ project that calls into a wasm module built in rust aswell ? but off hand I dont know how to set that up.

**Reply by kripken** [reply]
> Is there a minimal example confirming that it does actually work in C++?

Threads, do you mean? Or something more specific?

For threads, there are lots of minimal tests in the test suite, grep for `test_pthread_create` for example.
